pkg/proc/i386_arch.go

```diff
@@ -0,0 +1,369 @@
+	r := make(map[string]int)
+	for regNum, regName := range i386DwarfToName {
+		r[strings.ToLower(regName)] = regNum
+	}
+	r["eflags"] = 9
+	r["st0"] = 11
+	r["st1"] = 12
+	r["st2"] = 13
+	r["st3"] = 14
+	r["st4"] = 15
+	r["st5"] = 16
+	r["st6"] = 17
+	r["st7"] = 18
+	return r
+}()
+
+func maxI386DwarfRegister() int {
+	max := int(i386DwarfIPRegNum)
+	for i := range i386DwarfToName {
+		if i > max {
+			max = i
+		}
+	}
+	return max
+}
+
+func (i *I386) RegistersToDwarfRegisters(staticBase uint64, regs Registers) op.DwarfRegisters {
+	dregs := make([]*op.DwarfRegister, maxI386DwarfRegister()+1)
+
+	for _, reg := range regs.Slice(true) {
+		if dwarfReg, ok := i386NameToDwarf[strings.ToLower(reg.Name)]; ok {
+			dregs[dwarfReg] = reg.Reg
+		}
+	}
+
+	return op.DwarfRegisters{
+		StaticBase: staticBase,
+		Regs:       dregs,
+		ByteOrder:  binary.LittleEndian,
+		PCRegNum:   i386DwarfIPRegNum,
+		SPRegNum:   i386DwarfSPRegNum,
+		BPRegNum:   i386DwarfBPRegNum,
+	}
+}
+
+// AddrAndStackRegsToDwarfRegisters returns DWARF registers from the passed in
+// PC, SP, and BP registers in the format used by the DWARF expression interpreter.
+func (i *I386) AddrAndStackRegsToDwarfRegisters(staticBase, pc, sp, bp, lr uint64) op.DwarfRegisters {
+	dregs := make([]*op.DwarfRegister, i386DwarfIPRegNum+1)
+	dregs[i386DwarfIPRegNum] = op.DwarfRegisterFromUint64(pc)
+	dregs[i386DwarfSPRegNum] = op.DwarfRegisterFromUint64(sp)
+	dregs[i386DwarfBPRegNum] = op.DwarfRegisterFromUint64(bp)
+
+	return op.DwarfRegisters{
+		StaticBase: staticBase,
+		Regs:       dregs,
+		ByteOrder:  binary.LittleEndian,
+		PCRegNum:   i386DwarfIPRegNum,
+		SPRegNum:   i386DwarfSPRegNum,
+		BPRegNum:   i386DwarfBPRegNum,
+	}
+}
+
+func (i *I386) DwarfRegisterToString(j int, reg *op.DwarfRegister) (name string, floatingPoint bool, repr string) {
+	name, ok := i386DwarfToName[j]
+	if !ok {
+		name = fmt.Sprintf("unknown%d", j)
+	}
+
+	switch n := strings.ToLower(name); n {
+	case "eflags":
+		return name, false, eflagsDescription.Describe(reg.Uint64Val, 32)
+
+	case "tw", "fop":
+		return name, true, fmt.Sprintf("%#04x", reg.Uint64Val)
+
+	default:
+		if reg.Bytes != nil && strings.HasPrefix(name, "xmm") {
+			return name, true, formatSSEReg(reg.Bytes)
+		} else if reg.Bytes != nil && strings.HasPrefix(name, "st(") {
+			return name, true, formatX87Reg(reg.Bytes)
+		} else if reg.Bytes == nil || (reg.Bytes != nil && len(reg.Bytes) <= 8) {
+			return name, false, fmt.Sprintf("%#016x", reg.Uint64Val)
+		} else {
+			return name, false, fmt.Sprintf("%#x", reg.Bytes)
+		}
+	}
+}
+
+// InhibitStepInto returns whether StepBreakpoint can be set at pc.
+// When cgo or pie on 386 linux, compiler will insert more instructions (ex: call __x86.get_pc_thunk.).
+// StepBreakpoint shouldn't be set on __x86.get_pc_thunk and skip it.
+// See comments on stacksplit in $GOROOT/src/cmd/internal/obj/x86/obj6.go for generated instructions details.
+func (i *I386) InhibitStepInto(bi *BinaryInfo, pc uint64) bool {
```

Comment:
cc @derekparker what if instead of doing this and having to read the symbol table we just didn't step into functions for which we don't have a debug_info entry? We could make a note about this, patch the linker to emit DIEs for __x86.get_pc_thunk and then use that in a couple of years.

Reply:
Need a decision

Reply:
Sorry for the delay, I didn't see the ping on this. I think I'm alright with this solution as is.